cdap-data-fabric/src/test/java/co/cask/cdap/api/dataset/lib/SystemDatasetDefinitionTest.java

```diff
@@ -0,0 +1,284 @@
+      Assert.fail("reconfigure should have thrown exception");
+    } catch (IncompatibleUpdateException e) {
+      // expected
+    }
+
+    DatasetProperties longIprops = PartitionedFileSetProperties.builder().setPartitioning(
+      Partitioning.builder().addLongField("i").addStringField("s").build()).build();
+    try {
+      ((Reconfigurable) pfsDef).reconfigure("pfs", longIprops, spec);
+      Assert.fail("reconfigure should have thrown exception");
+    } catch (IncompatibleUpdateException e) {
+      // expected
+    }
+
+    DatasetProperties revProps = PartitionedFileSetProperties.builder().setPartitioning(
+      Partitioning.builder().addStringField("s").addIntField("i").build()).build();
+    try {
+      ((Reconfigurable) pfsDef).reconfigure("pfs", revProps, spec);
+      Assert.fail("reconfigure should have thrown exception");
+    } catch (IncompatibleUpdateException e) {
+      // expected
+    }
+  }
+
+  @Test
+  public void testIndexedTableReconfigure() throws IncompatibleUpdateException {
+    DatasetDefinition indexedTableDef = registry.get(IndexedTable.class.getName());
+    Assert.assertTrue(indexedTableDef instanceof Reconfigurable);
+
+    DatasetProperties props = DatasetProperties.builder()
+      .add(IndexedTable.INDEX_COLUMNS_CONF_KEY, "a,b,c")
+      .add(Table.PROPERTY_READLESS_INCREMENT, "false")
+      .build();
+    DatasetSpecification spec = indexedTableDef.configure("idxtb", props);
+
+    DatasetProperties compat = DatasetProperties.builder()
+      .add(IndexedTable.INDEX_COLUMNS_CONF_KEY, "c,b,a")
+      .add(Table.PROPERTY_READLESS_INCREMENT, "true") // turning on is ok
+      .build();
+    spec = ((Reconfigurable) indexedTableDef).reconfigure("idxtb", compat, spec);
+
+    DatasetProperties incompat = DatasetProperties.builder()
+      .add(IndexedTable.INDEX_COLUMNS_CONF_KEY, "a,d")
+      .add(Table.PROPERTY_READLESS_INCREMENT, "true")
+      .build();
+    try {
+      ((Reconfigurable) indexedTableDef).reconfigure("idxtb", incompat, spec);
+      Assert.fail("reconfigure should have thrown exception");
+    } catch (IncompatibleUpdateException e) {
+      // expected
+    }
+
+    incompat = DatasetProperties.builder()
+      .add(IndexedTable.INDEX_COLUMNS_CONF_KEY, "a,b,c")
+      .add(Table.PROPERTY_READLESS_INCREMENT, "false") // tirning off is not ok
+      .build();
+    try {
+      ((Reconfigurable) indexedTableDef).reconfigure("idxtb", incompat, spec);
+      Assert.fail("reconfigure should have thrown exception");
+    } catch (IncompatibleUpdateException e) {
+      // expected
+    }
+  }
+}
+
+class DatasetDefinitionRegistryWithDefaultModules extends DefaultDatasetDefinitionRegistry {
+
+  @Inject
+  DatasetDefinitionRegistryWithDefaultModules(Injector injector,
+                                              @Named("defaultDatasetModules")
+                                                Map<String, DatasetModule> defaultModules) {
```

Comment:
It looks like that, but because it is a continuation of the previous line, IntelliJ indents it this way.